### PR TITLE
[alt] typing: accept buffers in `IO.write`

### LINKED ISSUE #9082: Support buffers in `IO[bytes].write`
Trying to annotate psycopg code in order to pass mypy check with disabled bytearray/memoryview promotion to bytes [see branch mypy-0990](https://github.com/psycopg/psycopg/tree/mypy-0990).

Everything was smooth, except one questionable point:

```python
from typing import IO, Union
from typing_extensions import TypeAlias

Buffer: TypeAlias = Union[bytes, bytearray, memoryview]

def write1(f: IO[bytes], data: Buffer) -> None:
    f.write(data)
```

Mypy 0.990 reports:

```
error: Argument 1 to "write" of "IO" has incompatible type "Union[bytes, bytearray, memoryview]"; expected "bytes"  [arg-type]
```

Same using `BinaryIO` instead of `IO[bytes]`.

As far as I can see, a binary file supports any object supporting the buffer protocol (or maybe less than that? However, the three types in the union tested are definitely supported).

Declaring `f: IO[Buffer]` almost works: `write()` doesn't complain anymore, but `IO` does: *Invalid type argument value for "IO"*.

Maybe something could be relaxed here?

### PR DESCRIPTION
**Pull request by AlexWaygood**
An alternative PR to #9084 (closes #9084)

Fixes #9082

Co-authored-by: JelleZijlstra <jelle.zijlstra@gmail.com>

### CODE REVIEW

File: stdlib/tempfile.pyi [hunk 1]
```diff
@@ -215,7 +215,17 @@ class _TemporaryFileWrapper(Generic[AnyStr], IO[AnyStr]):
...
+    @overload
+    def write(self: _TemporaryFileWrapper[bytes], s: ReadableBuffer) -> int: ...
+    @overload
     def write(self, s: AnyStr) -> int: ...
+    @overload
+    def writelines(self: _TemporaryFileWrapper[str], lines: Iterable[str]) -> None: ...
+    @overload
+    def writelines(self: _TemporaryFileWrapper[bytes], lines: Iterable[ReadableBuffer]) -> None: ...
+    @overload
     def writelines(self, lines: Iterable[AnyStr]) -> None: ...
```
Comment (srittau) [inline stdlib/tempfile.pyi:229]: What's the purpose of the overloads here? Why not use a simple union?

File: stdlib/tempfile.pyi [hunk 2]
```diff
@@ -215,7 +215,17 @@ class _TemporaryFileWrapper(Generic[AnyStr], IO[AnyStr]):
     def tell(self) -> int: ...
     def truncate(self, size: int | None = ...) -> int: ...
     def writable(self) -> bool: ...
+    @overload
+    def write(self: _TemporaryFileWrapper[str], s: str) -> int: ...
+    @overload
+    def write(self: _TemporaryFileWrapper[bytes], s: ReadableBuffer) -> int: ...
+    @overload
     def write(self, s: AnyStr) -> int: ...
```
Comment (srittau) [inline stdlib/tempfile.pyi:223]: Is the third overload needed? This should be covered by the preceding two overloads. (Same for further overloads below.)

Comment (AlexWaygood) [reply stdlib/tempfile.pyi:223]: Unfortunately the test cases I've added don't pass for pyright unless we include this third overload. This is due to a design decision by pyright in the way it handles contrained `TypeVar`s that leads to different behaviour from mypy: https://github.com/microsoft/pyright/blob/main/docs/mypy-comparison.md#constrained-type-variables

It's the same issue that led to https://github.com/microsoft/pyright/issues/4534 being filed, which we then fixed over at typeshed in #9592.

**Comment by AlexWaygood** [comment]
> Diff from [mypy_primer](https://github.com/hauntsaninja/mypy_primer), showing the effect of this PR on open source code:

Two errors go away (`psycopg`, `cwltool`), one new error appears (`rich`), and two existing errors have their error code changed (`urllib3`). Looks promising to me. @JelleZijlstra, what do you think?

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**